pkg/server/devfile-handler.go

```diff
@@ -0,0 +1,181 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	buildv1 "github.com/openshift/api/build/v1"
+	imagev1 "github.com/openshift/api/image/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	"github.com/openshift/console/pkg/serverutils"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/klog"
+
+	devfilev1 "github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+	devfile "github.com/devfile/library/pkg/devfile"
+	"github.com/devfile/library/pkg/devfile/generator"
+	"github.com/devfile/library/pkg/devfile/parser"
+)
+
+const (
+	dockerfilePath = "Dockerfile"
+)
+
+func (s *Server) devfileHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		data       devfileForm
+		devfileObj parser.DevfileObj
+	)
+
+	err := json.NewDecoder(r.Body).Decode(&data)
+	if err != nil {
+		errMsg := fmt.Sprintf("Failed to decode response: %v", err)
+		klog.Error(errMsg)
+		serverutils.SendResponse(w, http.StatusInternalServerError, serverutils.ApiError{Err: errMsg})
+		return
+	}
+
+	// Get devfile content and parse it using a library call in the future
+	devfileContentBytes := []byte(data.Devfile.DevfileContent)
+	devfileObj, err = devfile.ParseFromDataAndValidate(devfileContentBytes)
+	if err != nil {
+		errMsg := fmt.Sprintf("Failed to parse devfile: %v", err)
+		klog.Error(errMsg)
+		serverutils.SendResponse(w, http.StatusInternalServerError, serverutils.ApiError{Err: errMsg})
+		return
+	}
+
+	containerComponents := devfileObj.Data.GetDevfileContainerComponents() //TODO: filter thru Console attribute, right now if there is more than one component container err out
+	if len(containerComponents) != 1 {
+		errMsg := fmt.Sprintf("Console Devfile Import Dev Preview, supports only one component container")
+		klog.Error(errMsg)
+		serverutils.SendResponse(w, http.StatusInternalServerError, serverutils.ApiError{Err: errMsg})
+		return
+	}
+
+	deploymentResource, err := getDeployResource(data, devfileObj)
+	if err != nil {
+		errMsg := fmt.Sprintf("Failed to get deployment resource for the devfile: %v", err)
+		klog.Error(errMsg)
+		serverutils.SendResponse(w, http.StatusInternalServerError, serverutils.ApiError{Err: errMsg})
+		return
+	}
+
+	service, err := getService(devfileObj)
+	if err != nil {
+		errMsg := fmt.Sprintf("Failed to get service for the devfile: %v", err)
+		klog.Error(errMsg)
+		serverutils.SendResponse(w, http.StatusInternalServerError, serverutils.ApiError{Err: errMsg})
+		return
+	}
+
+	devfileResources := devfileResources{
+		ImageStream:    getImageStream(),
+		BuildResource:  getBuildResource(data),
+		DeployResource: deploymentResource,
+		Service:        service,
+		Route:          getRoutes(data, containerComponents),
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	resp, _ := json.Marshal(devfileResources)
```

Comment:
We should be logging any errors.

Reply:
Odd, Jakub's code... happy to do it though. I'll log it as a 500 Internal Server Error as if it failed to marshal the json, it's beyond the caller to manage